src/intelligence_layer/connectors/retrievers/hybrid_qdrant_in_memory_retriever.py

```diff
@@ -0,0 +1,206 @@
+        sparse_embedder = self._search_client._get_or_init_sparse_model(
+            model_name=self._search_client.sparse_embedding_model_name
+        )
+
+        sparse_vector = next(sparse_embedder.query_embed(query=query))
+        sparse_query_vector = models.SparseVector(
+            indices=sparse_vector.indices.tolist(),
+            values=sparse_vector.values.tolist(),
+        )
+
+        return sparse_query_vector
+
+    def get_relevant_documents_with_scores(
+        self, query: str, filter: Optional[models.Filter] = None
+    ) -> Sequence[SearchResult[int]]:
+        """Search for documents in memory using hybrid search.
+
+        This method first embeds the given query using both a dense and sparse embedding model.
+        Then, it performs two search requests, one for each embedding type.
+        The results are then fused using the Reciprocal Rank Fusion algorithm.
+
+        Args:
+            query: The text to be searched with.
+            filter: Optional conditions to filter by.
+
+        Returns:
+            All documents that correspond to the query and pass the filter,
+            sorted by their reciprocal rank fusion score.
+        """
+        dense_query_vector = self._embed(query, self._query_representation)
+        sparse_query_vector = self._sparse_embed_query(query)
+
+        dense_request = models.SearchRequest(
+            vector=models.NamedVector(
+                name=self._dense_vector_field_name,
+                vector=dense_query_vector,
+            ),
+            limit=self._k,
+            filter=filter,
+            with_payload=True,
+            score_threshold=self._threshold,
+        )
+        sparse_request = models.SearchRequest(
+            vector=models.NamedSparseVector(
+                name=self._sparse_vector_field_name,
+                vector=sparse_query_vector,
+            ),
+            limit=self._k,
+            filter=filter,
+            with_payload=True,
+        )
+
+        dense_request_response, sparse_request_response = (
+            self._search_client.search_batch(
+                collection_name=self._collection_name,
+                requests=[dense_request, sparse_request],
+            )
+        )
+        search_result = reciprocal_rank_fusion(
+            [dense_request_response, sparse_request_response], limit=self._k
+        )
+
+        return [self._point_to_search_result(point) for point in search_result]
+
+    def get_filtered_documents_with_scores(
+        self, query: str, filter: models.Filter
+    ) -> Sequence[SearchResult[int]]:
+        """Specific method for `InMemoryRetriever` to support filtering search results.
+
+        Args:
+            query: The text to be searched with.
+            filter: Conditions to filter by.
+
+        Returns:
+            All documents that correspond to the query and pass the filter.
+        """
+        return self.get_relevant_documents_with_scores(query, filter)
```

Comment:
would it not have made more sense to implement these two functions the other way around, implement this and let the other call it? This way you are modifying the existing signature of `get_relevant_documents_with_scores`

Reply:
Good point, I will update the code like this! :+1: